aea/cli/test.py

```diff
@@ -0,0 +1,141 @@
+@click.pass_context
+@check_aea_project
+@pytest_args
+def test(click_context: click.Context, args: Sequence[str]) -> None:
+    """Run tests of an AEA project."""
+    click.echo("Executing tests of the AEA project...")
+    ctx = cast(Context, click_context.obj)
+    if click_context.invoked_subcommand is None:
+        test_package_by_path(Path(ctx.cwd), args)
+
+
+@test.command()
+@click.argument("connection_public_id", type=PublicIdParameter(), required=True)
+@pytest_args
+@pass_ctx
+def connection(
+    ctx: Context, connection_public_id: PublicId, args: Sequence[str]
+) -> None:
+    """Executes a test suite of a connection package dependency."""
+    test_item(ctx, CONNECTION, connection_public_id, args)
+
+
+@test.command()
+@click.argument("contract_public_id", type=PublicIdParameter(), required=True)
+@pytest_args
+@pass_ctx
+def contract(ctx: Context, contract_public_id: PublicId, args: Sequence[str]) -> None:
+    """Executes a test suite of a contract package dependency."""
+    test_item(ctx, CONTRACT, contract_public_id, args)
+
+
+@test.command()
+@click.argument("protocol_public_id", type=PublicIdParameter(), required=True)
+@pytest_args
+@pass_ctx
+def protocol(ctx: Context, protocol_public_id: PublicId, args: Sequence[str]) -> None:
+    """Executes a test suite of a protocol package dependency."""
+    test_item(ctx, PROTOCOL, protocol_public_id, args)
+
+
+@test.command()
+@click.argument("skill_public_id", type=PublicIdParameter(), required=True)
+@pytest_args
+@pass_ctx
+def skill(ctx: Context, skill_public_id: PublicId, args: Sequence[str]) -> None:
+    """Executes a test suite of a skill package dependency."""
+    test_item(ctx, SKILL, skill_public_id, args)
+
+
+@test.command()
+@click.argument(
+    "path", type=click.Path(exists=True, file_okay=False, dir_okay=True), required=True
+)
+@pytest_args
+@pass_ctx
+def by_path(
+    ctx: Context,
+    path: str,
+    args: Sequence[str],
+) -> None:
+    """Executes a test suite of a package specified by a path."""
+    click.echo(f"Executing tests of package at {path}'...")
+    full_path = Path(ctx.cwd) / Path(path)
+    test_package_by_path(full_path, args)
+
+
+def test_item(
+    ctx: Context,
+    item_type: str,
+    item_public_id: PublicId,
+    pytest_arguments: Sequence[str],
+) -> None:
+    """
+    Run tests of a package dependency.
+
+    :param ctx: the context.
+    :param item_type: the item type.
+    :param item_public_id: the item public id.
+    :param pytest_arguments: arguments to forward to Pytest
+    """
+    item_type_plural = item_type + "s"
+    click.echo(
+        "Executing tests of component of type {}, {}' ...".format(
+            item_type, item_public_id
+        )
+    )
+    package_dirpath = Path(ctx.cwd, item_type_plural, item_public_id.name)
+    test_package_by_path(package_dirpath, pytest_arguments)
+
+
+def test_package_by_path(package_dir: Path, pytest_arguments: Sequence[str]) -> None:
```

Comment:
all three modes reduce to call this function, exploiting the ability of `pytest` to collect tests recursively (to be tested). By convention, package tests are in a direct subdirectory `tests/` (but not enforced yet, still TODO)